src/internal_modules/roc_core/limited_arena.cpp

```diff
@@ -0,0 +1,42 @@
+/*
+ * Copyright (c) 2023 Roc Streaming authors
+ *
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+
+#include "roc_core/limited_arena.h"
+
+namespace roc {
+namespace core {
+
+LimitedArena::LimitedArena(IArena& arena, MemoryLimiter& memory_limiter)
+    : arena_(arena)
+    , memory_limiter_(memory_limiter) {
+}
+
+void* LimitedArena::allocate(size_t size) {
+    size_t allocated_size = arena_.compute_allocated_size(size);
+    if (memory_limiter_.acquire(allocated_size)) {
+        return arena_.allocate(size);
```

Comment:
If nested arena returns NULL, we should release acquired memory.

Reply:
Good point. For sure.

Reply:
It would be hard to test this, unless I build a mock IArena that returns nullptr. Let me know if you desire this.

Reply:
Up to you, 100% test coverage is not a requirement. It would be fine without a test too.